apollo-router/src/router_factory.rs

```diff
@@ -724,6 +736,105 @@ fn inject_schema_id(schema_id: Option<&str>, configuration: &mut Value) {
     }
 }
 
+// The Rust QP has not yet implemented setContext
+// (`@context` directives), progressive overrides, and it
+// doesn't support fed v1 *supergraphs*.
+//
+// If users are using the Rust QP as standalone (`new`) or in comparison mode (`both`),
+// fail to start up the router emitting an error.
+fn can_use_with_experimental_query_planner(
+    configuration: Arc<Configuration>,
+    schema: Arc<Schema>,
+) -> Result<(), ConfigurationError> {
+    match configuration.experimental_query_planner_mode {
+        crate::configuration::QueryPlannerMode::New
+        | crate::configuration::QueryPlannerMode::Both => {
+            // We have a *progressive* override when `join__directive` has a
+            // non-null value for `overrideLabel` field.
+            //
+            // This looks at object types' fields and their directive
+            // applications, looking specifically for `@join__direcitve`
+            // arguments list.
+            let has_progressive_overrides = schema
+                .supergraph_schema()
+                .types
+                .values()
+                .filter_map(|extended_type| {
+                    // The override label args can be only on ObjectTypes
+                    if let ExtendedType::Object(object_type) = extended_type {
+                        Some(object_type)
+                    } else {
+                        None
+                    }
+                })
+                .flat_map(|object_type| &object_type.fields)
+                .filter_map(|(_, field)| {
+                    let join_field_directives = field
+                        .directives
+                        .iter()
+                        .filter(|d| d.name.as_str() == JOIN_DIRECTIVE)
+                        .collect::<Vec<_>>();
+                    if !join_field_directives.is_empty() {
+                        Some(join_field_directives)
+                    } else {
+                        None
+                    }
+                })
+                .flatten()
+                .any(|join_directive| {
+                    if let Some(override_label_arg) =
+                        join_directive.argument_by_name(OVERRIDE_LABEL_ARG_NAME)
+                    {
+                        if !override_label_arg.is_null() {
+                            return true;
+                        }
+                        return false;
+                    }
+                    false
+                });
+            if has_progressive_overrides {
+                return Err(ConfigurationError::InvalidConfiguration {
+                    message: "`experimental_query_planner_mode` cannot be used with progressive overrides",
+                    error: "remove uses of progressive overrides to try the experimental_query_planner_mode in `both` or `new`, otherwise switch back to `legacy`.".to_string(),
+                });
```

Comment:
Depending on how you look at it, it might not be a `InvalidConfiguration` error. I didn't want to add another variant to the _public_ `ConfigurationError` enum since this is really temporary. Perhaps there is a better suited error type we've got that I can't think of?